File: implement-shell-tools/cat/cat.js
import { program } from "commander";
import { promises as fs } from "node:fs";
import process from "node:process";

program
    .name("cat")
    .description("Reading the file and print the content")
    .option("-n", "Number all lines")
    .option("-b", "Number only lines with content")
    .argument("<path...>", "The file path to process");

program.parse();

const paths = program.args;
const flags = program.opts();
// console.log(flags)

for (const path of paths) {
    try {
        const content = await fs.readFile(path, "utf-8");
        const lines = content.split("\n");
        lines.pop()
        // console.log("lines", lines)
        let lineNum = 1;

        for (let i = 0; i < lines.length; i++) {
            const line = lines[i];
            if (flags.n) {
                console.log(`${lineNum}  ${line}`)
                lineNum++
            } else if (flags.b) {
                if (line != "") {
                    console.log(`${lineNum} ${line}`)
                    lineNum++
                }
            }
            else {
                console.log(line)
            }
        }
    } catch (err) {
        console.error(`${path}: ${err.message}`)
    }
}

File: implement-shell-tools/cat/cat.py
import sys
import argparse

parser = argparse.ArgumentParser(
    prog="cat",
    description="Reading the file and print the content",
)

parser.add_argument("-n", action="store_true", help="Number all lines")
parser.add_argument("-b", action="store_true", help="Number only lines with content")
parser.add_argument("paths", nargs="+", help="The file path to process")

args = parser.parse_args()
paths = args.paths

for path in paths:
    try:
        with open(path, "r") as f:
            content = f.read()
        lines = content.splitlines()
        
        line_num = 1
        
        for i in range(len(lines)):
            line = lines[i]

            if args.b:
                if line != "":
                    print(f"{line_num} {line}")
                    line_num += 1
            elif args.n:
                print(f"{line_num} {line}")
                line_num += 1
            else:
                print(line)

    except Exception as e:
        print(f"Error reading {path}: {e}")